test/integration/array_test.ts

```diff
@@ -733,4 +733,498 @@ describe('Array', function () {
+        undefined,
+      );
+
+      assert.deepEqual(
+        doc.getRoot().listObjects.find((obj) => obj.id === 'first'),
+        { id: 'first' },
+      );
+
+      doc.getRoot().list.find(
+        function (this: any) {
+          assert.strictEqual(this.hello, 'world');
+          return true;
+        },
+        { hello: 'world' },
+      );
+    });
+
+    it('findIndex()', () => {
+      const doc = DocumentReplica.create<TestDoc>('test-doc');
+      doc.update((root) => {
+        root.empty = [];
+        root.list = [1, 2, 3];
+      });
+
+      assert.strictEqual(
+        doc.getRoot().empty.findIndex(() => true),
+        -1,
+      );
+
+      assert.strictEqual(
+        doc.getRoot().list.findIndex((num) => num >= 2),
+        1,
+      );
+
+      assert.strictEqual(
+        doc.getRoot().list.findIndex((num) => num >= 4),
+        -1,
+      );
+
+      doc.getRoot().list.findIndex(
+        function (this: any) {
+          assert.strictEqual(this.hello, 'world');
+          return true;
+        },
+        { hello: 'world' },
+      );
+    });
+
+    it('forEach()', () => {
+      const doc = DocumentReplica.create<TestDoc>('test-doc');
+      doc.update((root) => {
+        root.empty = [];
+        root.list = [1, 2, 3];
+      });
+
+      doc
+        .getRoot()
+        .empty.forEach(() =>
+          assert.fail('was called', 'not called', 'callback error'),
+        );
+
+      const testList: Array<number> = [];
+      doc.getRoot().list.forEach((num) => testList.push(num + 1));
+      assert.deepStrictEqual(testList, [2, 3, 4]);
+
+      doc.getRoot().list.forEach(
+        function (this: any) {
+          assert.strictEqual(this.hello, 'world');
+          return true;
+        },
+        { hello: 'world' },
+      );
+    });
+
+    it('includes()', () => {
+      const doc = DocumentReplica.create<TestDoc>('test-doc');
+      doc.update((root) => {
+        root.list = [1, 2, 3];
+      });
+
+      assert.strictEqual(doc.getRoot().list.includes(3), true);
+      assert.strictEqual(doc.getRoot().list.includes(0), false);
+      assert.strictEqual(doc.getRoot().list.includes(1, 1), false);
+      assert.strictEqual(doc.getRoot().list.includes(2, -2), true);
+    });
+
+    it('indexOf()', () => {
+      const doc = DocumentReplica.create<TestDoc>('test-doc');
+      doc.update((root) => {
+        root.list = [1, 2, 3];
+      });
+
+      assert.strictEqual(doc.getRoot().list.indexOf(3), 2);
+      assert.strictEqual(doc.getRoot().list.indexOf(0), -1);
+      assert.strictEqual(doc.getRoot().list.indexOf(1, 1), -1);
+      assert.strictEqual(doc.getRoot().list.indexOf(2, -2), 1);
+    });
+
+    it('indexOf() with objects', () => {
+      const doc = DocumentReplica.create<TestDoc>('test-doc');
+      doc.update((root) => {
+        root.listObjects = [{ id: 'first' }, { id: 'second' }];
+      });
+
+      // TODO: test always fails because doc.getRoot() returns a new proxy of cloned root.
```

Comment:
In the `indexOf` test code, `doc.getRoot()` always returns a new proxy, so I can't test it.

Reply:
🤔 I can't think of any good ideas for this right now. We may have to accept it as a limitation of ArrayProxy.